apps/obsidian/src/components/canvas/shapes/DiscourseNodeShape.tsx

```diff
@@ -0,0 +1,206 @@
+    return (
+      <HTMLContainer>
+        {createElement(discourseNodeContent, {
+          shape,
+          app: this.options.app,
+          canvasFile: this.options.canvasFile,
+          plugin: this.options.plugin,
+        })}
+      </HTMLContainer>
+    );
+  }
+
+  indicator(shape: DiscourseNodeShape) {
+    return <rect width={shape.props.w} height={shape.props.h} />;
+  }
+
+  async getFile(
+    shape: DiscourseNodeShape,
+    ctx?: { app: App; canvasFile: TFile },
+  ): Promise<TFile | null> {
+    const app = ctx?.app ?? this.options.app;
+    const canvasFile = ctx?.canvasFile ?? this.options.canvasFile;
+    return resolveLinkedFileFromSrc({
+      app,
+      canvasFile,
+      src: shape.props.src ?? undefined,
+    });
+  }
+
+  async getFrontmatter(
+    shape: DiscourseNodeShape,
+    ctx?: { app: App; canvasFile: TFile },
+  ): Promise<FrontmatterRecord | null> {
+    const app = ctx?.app ?? this.options.app;
+    const file = await this.getFile(shape, ctx);
+    if (!file) return null;
+    return getFrontmatterForFile(app, file);
+  }
+
+  async getDiscourseNodeType(
+    shape: DiscourseNodeShape,
+    ctx?: { app: App; canvasFile: TFile },
+  ): Promise<DiscourseNode | null> {
+    const frontmatter = await this.getFrontmatter(shape, ctx);
+    const nodeTypeId = getNodeTypeIdFromFrontmatter(frontmatter);
+    return getNodeTypeById(this.options.plugin, nodeTypeId);
+  }
+
+  async getRelations(
+    shape: DiscourseNodeShape,
+    ctx?: { app: App; canvasFile: TFile },
+  ): Promise<unknown[]> {
+    const frontmatter = await this.getFrontmatter(shape, ctx);
+    if (!frontmatter) return [];
+    // TODO: derive relations from frontmatter
+    return [];
+  }
+}
+
+const discourseNodeContent = memo(
+  ({
+    shape,
+    app,
+    canvasFile,
+    plugin,
+  }: {
+    shape: DiscourseNodeShape;
+    app: App;
+    canvasFile: TFile;
+    plugin: DiscourseGraphPlugin;
+  }) => {
+    const editor = useEditor();
+    const { src, title, nodeTypeId } = shape.props;
+    const nodeType = getNodeTypeById(plugin, nodeTypeId);
+
+    useEffect(() => {
+      const loadNodeData = async () => {
+        if (!src) {
+          editor.updateShape<DiscourseNodeShape>({
+            id: shape.id,
+            type: "discourse-node",
+            props: {
+              ...shape.props,
+              title: "(no source)",
+            },
+          });
+          return;
+        }
+
+        try {
+          const linkedFile = await resolveLinkedFileFromSrc({
+            app,
+            canvasFile,
+            src,
+          });
+
+          if (!linkedFile) {
+            editor.updateShape<DiscourseNodeShape>({
+              id: shape.id,
+              type: "discourse-node",
+              props: {
+                ...shape.props,
+                title: "(unlinked)",
+              },
+            });
+            return;
+          }
+
+          editor.updateShape<DiscourseNodeShape>({
```

Comment:
We should only update if title !== basename, otherwise we might flicker the shapes.